compiler/rustc_const_eval/src/check_consts/check.rs

```
@@ -591,7 +591,7 @@ impl<'tcx> Visitor<'tcx> for Checker<'_, 'tcx> {
                 if let AggregateKind::Coroutine(def_id, ..) = kind.as_ref()
                     && let Some(
                         coroutine_kind @ hir::CoroutineKind::Desugared(
-                            hir::CoroutineDesugaring::Async,
+                            hir::CoroutineDesugaring::Async | hir::CoroutineDesugaring::Gen,
```

This line causes the `const_gen_fn.rs` test to emit an extra error. Before the error message just said "functions cannot be both `const` and `gen`" but now it also says "`gen` bodys are not allowed in constant functions."

I think this makes the error messages worse (the existing error was sufficient), but it also seems more correct to me to also visit `Gen` coroutines here.

---

What I would expect here is that we complain on *all* coroutines, except for an allowlist of those that are permitted in const. But currently the code seems to use a blocklist, which is bad since it means new things are allowed by default.

Why do we allow `AsyncGen` here? That's almost surely wrong, isn't it?

---

When I made this reject all coroutines, I got new errors on the following test snippets.

```rust
const G = #[coroutine] || -> i32 { yield 0; return 1; };
```

This one was just testing error messages, so I don't think it particularly matters if const coroutines are supported or not.


The second one comes from `issue-53678-coroutine-and-const-fn.rs`. This one seems to explicitly want to support coroutines in const. It's marked as `check-pass`, so we don't know whether const eval would actually succeed on this though.

```rust
pub type CoroOnce<Y, R> = impl Coroutine<Yield = Y, Return = R>;

#[define_opaque(CoroOnce)]
pub const fn const_coroutine<Y, R>(yielding: Y, returning: R) -> CoroOnce<Y, R> {
    #[coroutine]
    move || {
        yield yielding;

        return returning;
    }
}

const FOO: CoroOnce<usize, usize> = const_coroutine(10, 100);
```

I think I'll change this line to an explicit match to make it clearer what the intended behavior is, rather than silently falling through if it doesn't match.

---

Sounds good. I had to adjust `Coroutine::status_in_item` but this seems like a better approach.

---

Please add a FIXME that coroutines should be moved to a different feature gate eventually.

---

Done!